---
title: "CHO Hypothalamus BulkRNAseq"
author: "Si Young Lee"
date: "2025-12-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r}
library(dplyr); library(tidyr); library(ggplot2); library(rstatix)
library(pheatmap); library(clusterProfiler); library(org.Mm.eg.db)
library(glmnet); library(VennDiagram); library(grid); library(writexl)
library(VennDiagram);library(forcats);library(ggrepel); library(DESeq2)
library(reshape2);library(tidyverse); library(AnnotationDbi); library(GO.db)
library(viridis); library(enrichplot)
```

```{r import read count matrix}
count_hypo <- gene_count
sample_info_hypo <- Sample_Information_Bulk_Hypothalamus_CHO
```

```{r}
out_dir_hypo <- "output hypothalamus"
.group_base_colors <- c(
  "Chow-HFHS" = "#000000",
  "Lactation HFHS-HFHS" = "#de5528",
  "Post-weaning  HFHS-HFHS" = "#208696"
)

group_aes <- get_group_aesthetics(names(.group_base_colors))
fill_map <- group_aes$fill
color_map <- group_aes$color

group_order <- c("Chow-HFHS","Lactation HFHS-HFHS","Post-weaning  HFHS-HFHS")
write_xlsx(count_hypo, "read_count_bulk_hypo.xlsx")
```


```{r PCA}
count_hypo <- as.data.frame(count_hypo)
count_hypo$gene_name <- make.unique(count_hypo$gene_name, sep = "_dup")
rownames(count_hypo) <- count_hypo$gene_name
# Identify sample columns using sample_info_hypo
sample_cols <- intersect(colnames(count_hypo), sample_info_hypo$sample_name)

sample_cols <- intersect(colnames(count_hypo), sample_info_hypo$sample_name)

# ensure numeric sample columns
count_hypo[ , sample_cols] <- lapply(count_hypo[ , sample_cols], function(x){
    as.numeric(as.character(x))
})

count_hypo_filtered <- count_hypo[
  rowSums(count_hypo[, sample_cols] == 0) == 0,
]

# Set gene names as rownames
rownames(count_hypo_filtered) <- count_hypo_filtered$gene_name
count_hypo_filtered <- count_hypo_filtered[ , !(colnames(count_hypo) == "gene_name")]

dds <- DESeqDataSetFromMatrix(
  countData = round(count_hypo_filtered),
  colData   = sample_info_hypo,
  design    = ~ group
)

vst_mat <- vst(dds, blind = TRUE)
pca_mat <- assay(vst_mat)
pca <- prcomp(t(pca_mat), scale. = TRUE)

pca_df <- data.frame(
  PC1   = pca$x[,1],
  PC2   = pca$x[,2],
  group = sample_info_hypo$group,
  sample = sample_info_hypo$sample_name
)

# Percent variance
var_expl <- round(100 * pca$sdev^2 / sum(pca$sdev^2), 2)

group_colors <- c(
  "Chow-HFHS" = "#000000",      
  "Post-weaning  HFHS-HFHS" = "#208696", 
  "Lactation HFHS-HFHS" = "#de5528"   
)

pca_hypothalamus <- ggplot(pca_df, aes(x = PC1, y = PC2, color = group)) +
  geom_point(size = 4) +
  stat_ellipse(aes(fill = group), geom = "polygon", alpha = 0.15, color = NA) +
  scale_color_manual(values = group_colors) +
  scale_fill_manual(values = group_colors) +
  labs(
    title = "PCA of Hypothalamus Samples",
    x = paste0("PC1 (", var_expl[1], "%)"),
    y = paste0("PC2 (", var_expl[2], "%)")
  ) +
  theme_bw(base_size = 14)

pca_hypothalamus
ggsave(file.path(out_dir,"PCA Bulk Hypothalamus.png"), pca_hypothalamus, height = 4, width = 7)
```
```{r analyze PC1 loadings}
pc1_loadings <- pca$rotation[, 1]

pc1_df <- data.frame(
  gene = names(pc1_loadings),
  loading = pc1_loadings
)
pc1_pos_genes <- pc1_df %>%
  arrange(desc(loading)) %>%
  slice_head(n = 500) %>%
  pull(gene)

pc1_neg_genes <- pc1_df %>%
  arrange(loading) %>%
  slice_head(n = 500) %>%
  pull(gene)
ego_pc1_pos <- enrichGO(
  gene = pc1_pos_genes,
  OrgDb = org.Mm.eg.db,
  keyType = "SYMBOL",
  ont = "BP",
  pAdjustMethod = "BH",
  readable = TRUE
)

ego_pc1_neg <- enrichGO(
  gene = pc1_neg_genes,
  OrgDb = org.Mm.eg.db,
  keyType = "SYMBOL",
  ont = "BP",
  pAdjustMethod = "BH",
  readable = TRUE
)
dotplot(ego_pc1_pos, showCategory = 30) +
  ggtitle("PC1-positive loadings")

dotplot(ego_pc1_neg, showCategory = 30) +
  ggtitle("PC1-negative loadings")

```

```{r gene trajectory plots}
dds0 <- DESeqDataSetFromMatrix(
  countData = round(count_hypo_filtered),
  colData   = sample_info_hypo,
  design    = ~ group
)

dds0 <- dds0[rowSums(counts(dds0)) > 10, ]
dds_lrt <- DESeq(dds0, test = "LRT", reduced = ~1)

res_anova <- results(dds_lrt)

res_anova_sig <- as.data.frame(res_anova) %>%
  rownames_to_column("gene") %>%
  filter(padj < 0.05)
dds_wald <- DESeq(dds0)

res_lact_vs_post <- results(
  dds_wald,
  contrast = c("group", "Lactation HFHS-HFHS", "Post-weaning  HFHS-HFHS")
)
res_lact_vs_post <- as.data.frame(res_lact_vs_post)
res_lact_vs_post_df <- res_lact_vs_post %>%
  tibble::rownames_to_column("gene")

res_lact_vs_post_df <- dplyr::select(
  res_lact_vs_post_df,
  gene,
  log2FoldChange
)

res_chow_vs_post <- results(
  dds_wald,
  contrast = c("group", "Chow-HFHS", "Post-weaning  HFHS-HFHS")
)


res_chow_vs_post <- as.data.frame(res_chow_vs_post)
res_chow_vs_post_df <- res_chow_vs_post %>%
  tibble::rownames_to_column("gene")

res_chow_vs_post_df <- dplyr::select(
  res_chow_vs_post_df,
  gene,
  log2FoldChange
)

res_lact_vs_post_df <- res_lact_vs_post
res_lact_vs_post_df$gene <- rownames(res_lact_vs_post_df)

res_lact_vs_post_df <- res_lact_vs_post_df[, c("gene", "log2FoldChange")]

res_chow_vs_post_df <- res_chow_vs_post
res_chow_vs_post_df$gene <- rownames(res_chow_vs_post_df)

res_chow_vs_post_df <- res_chow_vs_post_df[, c("gene", "log2FoldChange")]

res_anova_fc <- res_anova_sig %>%
  dplyr::select(gene, padj) %>%
  left_join(
    res_lact_vs_post_df %>%
      dplyr::rename(log2FC_Lact_vs_Post = log2FoldChange),
    by = "gene"
  ) %>%
  left_join(
    {
      df <- res_chow_vs_post
      df$gene <- rownames(df)
      df[, c("gene", "log2FoldChange")] %>%
        dplyr::rename(log2FC_Chow_vs_Post = log2FoldChange)
    },
    by = "gene"
  )


log2FC_cutoff <-0.00001 # adjust if needed

# Post-weaning upregulated (Post > both Chow and Lactation)
post_specific_up <- res_anova_fc[
  res_anova_fc$log2FC_Lact_vs_Post < -log2FC_cutoff &
  res_anova_fc$log2FC_Chow_vs_Post < -log2FC_cutoff, ]

# Post-weaning downregulated (Post < both Chow and Lactation)
post_specific_down <- res_anova_fc[
  res_anova_fc$log2FC_Lact_vs_Post > log2FC_cutoff &
  res_anova_fc$log2FC_Chow_vs_Post > log2FC_cutoff, ]

```

```{r t-test chow vs lactation}
dds <- DESeqDataSetFromMatrix(
    countData = round(count_hypo_filtered),
    colData   = sample_info_hypo,
    design    = ~ group
)

dds <- DESeq(dds)

res_lact_vs_chow <- results(
    dds,
    contrast = c("group", "Lactation HFHS-HFHS", "Chow-HFHS")
)

res_lact_vs_chow <- as.data.frame(res_lact_vs_chow)
sig_genes <- res_lact_vs_chow[res_lact_vs_chow$padj < 0.05, ]
nrow(sig_genes)
res_df_lact_chow <- as.data.frame(res_lact_vs_chow)
head(res_df_lact_chow)

res_df_lact_chow <- as.data.frame(res_lact_vs_chow)

deg_up_lact_chow_t <- res_df_lact_chow  %>%
  filter(padj < 0.05, log2FoldChange >0)

deg_down_lact_chow_t <- res_df_lact_chow %>%
  filter(padj < 0.05, log2FoldChange < -0.0000001)
```

```{r identigy genes that are only different in lactation compared to the other two}
# --- 1. Create DESeq2 dataset ---
dds <- DESeqDataSetFromMatrix(
  countData = round(count_hypo_filtered),   # round counts just in case
  colData = sample_info_hypo,
  design = ~ group
)

# Filter out low-count genes (optional but recommended)
dds <- dds[rowSums(counts(dds)) > 10, ]

# --- 2. Run DESeq2 ---
dds <- DESeq(dds)

# --- 3. Compute pairwise contrasts for Lactation ---
res_lact_vs_chow <- results(dds, contrast = c("group", "Lactation HFHS-HFHS", "Chow-HFHS"))
res_lact_vs_post <- results(dds, contrast = c("group", "Lactation HFHS-HFHS", "Post-weaning  HFHS-HFHS"))

# --- 4. Convert results to long format and add contrast labels ---
res_lact_vs_chow <- as.data.frame(res_lact_vs_chow) %>%
  rownames_to_column("gene") %>%
  mutate(contrast = "Lact_vs_Chow")

res_lact_vs_post <- as.data.frame(res_lact_vs_post) %>%
  rownames_to_column("gene") %>%
  mutate(contrast = "Lact_vs_Post")

res_long <- bind_rows(res_lact_vs_chow, res_lact_vs_post)

res_wide <- res_long %>%
  pivot_wider(
    id_cols = gene,                     # ensure gene is the unique identifier
    names_from = contrast, 
    values_from = c(log2FoldChange, padj)
  )


# --- 7. Identify Lactation-specific genes ---
log2FC_cutoff <- 0.00001
padj_cutoff <- 0.05

lactation_specific_up <- res_wide %>%
  filter(
    log2FoldChange_Lact_vs_Chow > log2FC_cutoff & padj_Lact_vs_Chow < padj_cutoff,
    log2FoldChange_Lact_vs_Post > log2FC_cutoff & padj_Lact_vs_Post < padj_cutoff
  )

lactation_specific_down <- res_wide %>%
  filter(
    log2FoldChange_Lact_vs_Chow < -log2FC_cutoff & padj_Lact_vs_Chow < padj_cutoff,
    log2FoldChange_Lact_vs_Post < -log2FC_cutoff & padj_Lact_vs_Post < padj_cutoff
  )

```

```{r ANOVA}
dds <- DESeqDataSetFromMatrix(
  countData = round(count_hypo_filtered),
  colData = sample_info_hypo,
  design = ~ group
)

dds <- dds[rowSums(counts(dds)) > 10, ]

dds <- DESeq(
  dds,
  test = "LRT",
  reduced = ~ 1
)
res_lrt <- results(dds)

res_lrt <- as.data.frame(res_lrt) %>%
  rownames_to_column("gene")

anova_sig_genes <- res_lrt %>%
  filter(padj < 0.05)
res_lact_vs_chow <- results(dds, contrast = c("group", "Lactation HFHS-HFHS", "Chow-HFHS"), test = "Wald")
res_lact_vs_post <- results(dds, contrast = c("group", "Lactation HFHS-HFHS", "Post-weaning  HFHS-HFHS"), test = "Wald")

res_fc <- as.data.frame(res_lact_vs_chow) %>%
  rownames_to_column("gene") %>%
  dplyr::select(gene, log2FoldChange) %>%
  dplyr::rename(log2FC_Lact_vs_Chow = log2FoldChange) %>%
  inner_join(
    as.data.frame(res_lact_vs_post) %>%
      rownames_to_column("gene") %>%
      dplyr::select(gene, log2FoldChange) %>%
      dplyr::rename(log2FC_Lact_vs_Post = log2FoldChange),
    by = "gene"
  )

res_anova_fc <- anova_sig_genes %>%
  inner_join(res_fc, by = "gene")

log2FC_cutoff <- 0.0000001

lactation_specific_up_anova <- res_anova_fc %>%
  filter(
    log2FC_Lact_vs_Chow > log2FC_cutoff,
    log2FC_Lact_vs_Post > log2FC_cutoff
  )

lactation_specific_down_anova <- res_anova_fc %>%
  filter(
    log2FC_Lact_vs_Chow < -log2FC_cutoff,
    log2FC_Lact_vs_Post < -log2FC_cutoff
  )

genes_up <- lactation_specific_up_anova$gene
genes_down <- lactation_specific_down_anova$gene
gene_universe <- res_anova_fc$gene

ego_up_anova <- enrichGO(
  gene          = genes_up,
  universe      = gene_universe,
  OrgDb         = org.Mm.eg.db,
  keyType       = "SYMBOL",
  ont           = "MF",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05,
  readable      = TRUE
)

ego_down_anova <- enrichGO(
  gene          = genes_down,
  universe      = gene_universe,
  OrgDb         = org.Mm.eg.db,
  keyType       = "SYMBOL",
  ont           = "MF",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05,
  readable      = TRUE
)

go_up_df <- as.data.frame(ego_up_anova)
go_down_df <- as.data.frame(ego_down_anova)
```

```{r lets compare up vs down 15% of the GO terms are related to protein folding, ER stress}
# UP genes
up_symbols <- rownames(deg_up_lact)
up_entrez <- bitr(
  up_symbols,
  fromType = "SYMBOL",
  toType = "ENTREZID",
  OrgDb = org.Mm.eg.db
)$ENTREZID

# DOWN genes
down_symbols <- rownames(deg_down_lact)
down_entrez <- bitr(
  down_symbols,
  fromType = "SYMBOL",
  toType = "ENTREZID",
  OrgDb = org.Mm.eg.db
)$ENTREZID

go_bp_up <- enrichGO(
  gene          = up_entrez,
  OrgDb         = org.Mm.eg.db,
  keyType       = "ENTREZID",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05,
  readable      = TRUE
)

go_bp_down <- enrichGO(
  gene          = down_entrez,
  OrgDb         = org.Mm.eg.db,
  keyType       = "ENTREZID",
  ont           = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05,
  readable      = TRUE
)

keywords <- c(
    "endoplasmic reticulum",
    "unfolded",
    "protein",
    "misfolded",
    "proteostasis",
    "ER stress"
)
go_bp_up_df <- as.data.frame(go_bp_up)
go_bp_down_df <- as.data.frame(go_bp_down)

top10_up <- go_bp_up_df %>%
  arrange(p.adjust) %>%
  slice_head(n = 10) %>%
  mutate(direction = "Upregulated")

top10_down <- go_bp_down_df %>%
  arrange(p.adjust) %>%
  slice_head(n = 10) %>%
  mutate(direction = "Downregulated")

go_top10_combined <- bind_rows(top10_up, top10_down)

top10_GO<- ggplot(go_top10_combined,
             aes(x = reorder(Description, -log10(p.adjust)),
                 y = -log10(p.adjust),
                 fill = direction)) +
        geom_col(width = 0.7) +
        coord_flip() +
        facet_wrap(~ direction, scales = "free_y") +
        scale_fill_manual(values = c(
          "Upregulated"   = "#C73E1D",
          "Downregulated" = "#2A7FFF"
        )) +
        labs(
          title = "Top 10 GO BP Terms in Up- and Downregulated Genes",
          x = NULL,
          y = expression(-log[10]("adjusted p-value")),
          fill = NULL
        ) +
        theme_minimal(base_size = 14) +
        theme(
          panel.grid.major.y = element_blank(),
          plot.title = element_text(face = "bold", hjust = 0.5)
        )

ggsave(
  filename = file.path(out_dir_hypo,"GO_enrichment_bp_up_down.png"),
  plot = top10_GO,
  width = 16,
  height = 8,
  dpi = 300
)

go_bp_up_focus <- go_bp_up_df %>%
  filter(grepl(paste(keywords, collapse = "|"),
               Description, ignore.case = TRUE))

go_bp_down_focus <- go_bp_down_df %>%
  filter(grepl(paste(keywords, collapse = "|"),
               Description, ignore.case = TRUE))

go_up_plot <- go_bp_up_focus %>%
  arrange(p.adjust) %>%
  mutate(
    Description = factor(Description, levels = rev(Description)),
    negLogP = -log10(p.adjust)
  )

go_down_plot <- go_bp_down_focus %>%
  arrange(p.adjust) %>%
  mutate(
    Description = factor(Description, levels = rev(Description)),
    negLogP = -log10(p.adjust)
  )
go_up_plot
col_up   <- "#C73E1D"   # muted red (stress)
col_down <- "#2A7FFF"   # soft blue (metabolic down)

go_up_plot$Direction   <- "Upregulated"
#go_down_plot$Direction <- "Downregulated"

go_up_plot$plot_value   <-  go_up_plot$negLogP
go_down_plot$plot_value <- -go_down_plot$negLogP

go_combined <- bind_rows(go_up_plot, go_down_plot)

ER_go <- ggplot(go_combined,
           aes(x = Description, y = plot_value, fill = Direction)) +
      geom_col(width = 0.75) +
      coord_flip() +
      scale_fill_manual(
        values = c("Upregulated" = col_up,
                   "Downregulated" = col_down)
      ) +
      geom_hline(yintercept = 0, color = "grey30", linewidth = 0.6) +
      labs(
        title = "Directional GO Enrichment in Hypothalamus",
        subtitle = "Lactation HFHS vs Chow HFHS",
        x = NULL,
        y = expression(-log[10]("adjusted p-value"))
      ) +
      theme_minimal(base_size = 14) +
      theme(
        plot.title = element_text(face = "bold", size = 16),
        plot.subtitle = element_text(size = 11, color = "grey40"),
        axis.text.y = element_text(size = 11),
        panel.grid.major.y = element_blank(),
        panel.grid.minor = element_blank(),
        legend.position = "top"
      )

ggsave(
  filename = "ER_stress_GO_enrichment_only.png",
  plot = ER_go,
  width = 12,
  height = 8,
  dpi = 300
)
```

```{r }
library(tidyverse)

go_bp_focus_df <- as.data.frame(go_bp_up_focus) %>%
  dplyr::select(Description, geneID) %>%
  separate_rows(geneID, sep = "/") %>%
  dplyr::rename(
    go_term = Description,
    gene = geneID
  )

res_df_lact_chow <- res_df_lact_chow %>%
  tibble::rownames_to_column(var = "gene")  # converts rownames to column 'gene'

go_bp_focus_df <- go_bp_focus_df %>%
  left_join(res_df_lact_chow %>% dplyr::select(gene, log2FoldChange), by = "gene")

circos.clear()
circos.par(
  start.degree = 90,
  gap.after = c(
    rep(2, length(unique(go_bp_focus_df$go_term)) - 1),
    8,
    rep(1, length(unique(go_bp_focus_df$gene)) - 1),
    8
  )
)

go_terms <- unique(go_bp_focus_df$go_term)
go_cols <- setNames(
  colorRampPalette(RColorBrewer::brewer.pal(8, "Set2"))(length(go_terms)),
  go_terms
)
gene_fc <- go_bp_focus_df %>%
  distinct(gene, log2FoldChange)

fc_col_fun <- colorRamp2(
  c(min(gene_fc$log2FoldChange),
    0,
    max(gene_fc$log2FoldChange)),
  c("#2A7FFF", "white", "#C73E1D")
)

gene_cols <- setNames(
  fc_col_fun(gene_fc$log2FoldChange),
  gene_fc$gene
)
chordDiagram(
  go_bp_focus_df %>% dplyr::select(go_term, gene),
  grid.col = c(go_cols, gene_cols),
  transparency = 0.4
)
circos.trackPlotRegion(
  track.index = 1,
  panel.fun = function(x, y) {
    circos.text(
      CELL_META$xcenter,
      CELL_META$ylim[1],
      CELL_META$sector.index,
      facing = "clockwise",
      niceFacing = TRUE,
      adj = c(0, 0.5),
      cex = 0.6
    )
  },
  bg.border = NA
)

```

```{r cc}
go_cc_up <- enrichGO(
  gene          = up_entrez,
  OrgDb         = org.Mm.eg.db,
  keyType       = "ENTREZID",
  ont           = "CC",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05,
  readable      = TRUE
)

go_cc_down <- enrichGO(
  gene          = down_entrez,
  OrgDb         = org.Mm.eg.db,
  keyType       = "ENTREZID",
  ont           = "CC",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05,
  readable      = TRUE
)

keywords <- c(
    "endoplasmic",
    "reticulum",
    "ribosome",
    "unfolded",
    "vesicle",
    "protein",
    "misfolded", 
    "transport", 
    "ribosomal", 
    "lysosomal"
)
go_cc_up_df <- as.data.frame(go_cc_up)
go_cc_down_df <- as.data.frame(go_cc_down)

top10_up <- go_cc_up_df %>%
  arrange(p.adjust) %>%
  slice_head(n = 10) %>%
  mutate(direction = "Upregulated")

top10_down <- go_cc_down_df %>%
  arrange(p.adjust) %>%
  slice_head(n = 10) %>%
  mutate(direction = "Downregulated")

go_top10_combined <- bind_rows(top10_up, top10_down)

top10_GO_cc<- ggplot(go_top10_combined,
             aes(x = reorder(Description, -log10(p.adjust)),
                 y = -log10(p.adjust),
                 fill = direction)) +
        geom_col(width = 0.7) +
        coord_flip() +
        facet_wrap(~ direction, scales = "free_y") +
        scale_fill_manual(values = c(
          "Upregulated"   = "#C73E1D",
          "Downregulated" = "#2A7FFF"
        )) +
        labs(
          title = "Top 10 GO CC Terms in Up- and Downregulated Genes",
          x = NULL,
          y = expression(-log[10]("adjusted p-value")),
          fill = NULL
        ) +
        theme_minimal(base_size = 14) +
        theme(
          panel.grid.major.y = element_blank(),
          plot.title = element_text(face = "bold", hjust = 0.5)
        )

ggsave(
  filename = file.path(out_dir_hypo, "GO_enrichment_cc_up_down.png"),
  plot = top10_GO_cc,
  width = 16,
  height = 8,
  dpi = 300
)


go_cc_up_focus <- go_cc_up_df %>%
  filter(grepl(paste(keywords, collapse = "|"),
               Description, ignore.case = TRUE))

go_cc_down_focus <- go_cc_down_df %>%
  filter(grepl(paste(keywords, collapse = "|"),
               Description, ignore.case = TRUE))

n_focus <- nrow(go_cc_up_focus)
n_total <- nrow(go_cc_up_df)

prop_label <- paste0(
  n_focus, " / ", n_total,
  " GO CC terms\n(~",
  round(100 * n_focus / n_total), "%)"
)

go_cc_up_focus %>%
  mutate(
    neglog10_padj = -log10(p.adjust),
    Description = reorder(Description, neglog10_padj)
  ) %>%
  ggplot(aes(
    x = neglog10_padj,
    y = Description,
    size = Count
  )) +
  geom_point(color = "#C73E1D", alpha = 0.85) +
  labs(
    title = "ER Stress / Protein Folding–Related GO CC Terms",
    subtitle = prop_label,
    x = expression(-log[10]("adjusted p-value")),
    y = NULL,
    size = "Gene count"
  ) +
  theme_minimal(base_size = 13)


```

```{r mf}
go_mf_up <- enrichGO(
  gene          = up_entrez,
  OrgDb         = org.Mm.eg.db,
  keyType       = "ENTREZID",
  ont           = "MF",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05,
  readable      = TRUE
)

go_mf_down <- enrichGO(
  gene          = down_entrez,
  OrgDb         = org.Mm.eg.db,
  keyType       = "ENTREZID",
  ont           = "MF",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05,
  readable      = TRUE
)

keywords <- c(
    "endoplasmic",
    "reticulum",
    "ribosome",
    "unfolded",
    "vesicle",
    "protein",
    "misfolded", 
    "transport", 
    "ribosomal", 
    "lysosomal"
)

go_mf_up_df <- as.data.frame(go_mf_up)
go_mf_down_df <- as.data.frame(go_mf_down)

top10_up <- go_mf_up_df %>%
  arrange(p.adjust) %>%
  slice_head(n = 10) %>%
  mutate(direction = "Upregulated")

top10_down <- go_mf_down_df %>%
  arrange(p.adjust) %>%
  slice_head(n = 10) %>%
  mutate(direction = "Downregulated")

go_top10_combined <- bind_rows(top10_up, top10_down)

top10_GO_mf<- ggplot(go_top10_combined,
             aes(x = reorder(Description, -log10(p.adjust)),
                 y = -log10(p.adjust),
                 fill = direction)) +
        geom_col(width = 0.7) +
        coord_flip() +
        facet_wrap(~ direction, scales = "free_y") +
        scale_fill_manual(values = c(
          "Upregulated"   = "#C73E1D",
          "Downregulated" = "#2A7FFF"
        )) +
        labs(
          title = "Top 10 GO MF Terms in Up- and Downregulated Genes",
          x = NULL,
          y = expression(-log[10]("adjusted p-value")),
          fill = NULL
        ) +
        theme_minimal(base_size = 14) +
        theme(
          panel.grid.major.y = element_blank(),
          plot.title = element_text(face = "bold", hjust = 0.5)
        )

ggsave(
  filename = file.path(out_dir_hypo, "GO_enrichment_mf_up_down.png"),
  plot = top10_GO_mf,
  width = 16,
  height = 8,
  dpi = 300
)

go_mf_up_focus <- go_mf_up_df %>%
  filter(grepl(paste(keywords, collapse = "|"),
               Description, ignore.case = TRUE))

go_mf_down_focus <- go_mf_down_df %>%
  filter(grepl(paste(keywords, collapse = "|"),
               Description, ignore.case = TRUE))

```


```{r genes driving GO}
library(tidyr)
library(dplyr)

go_genes_up <- go_bp_up_focus %>%
  dplyr::select(ID, Description, geneID, p.adjust) %>%
  mutate(
    geneID = strsplit(as.character(geneID), "/"),
    Direction = "Upregulated"
  ) %>%
  unnest(geneID)

go_genes_down <- go_bp_down_focus %>%
  dplyr::select(ID, Description, geneID, p.adjust) %>%
  mutate(
    geneID = strsplit(as.character(geneID), "/"),
    Direction = "Downregulated"
  ) %>%
  unnest(geneID)

go_genes_long <- bind_rows(go_genes_up, go_genes_down)
str(go_genes_long$geneID)

core_genes <- go_genes_long %>%
  dplyr::count(Direction, geneID, sort = TRUE)

head(core_genes, 15)

ggplot(core_genes %>% filter(n >= 2),
       aes(x = reorder(geneID, n), y = n, fill = Direction)) +
  geom_col() +
  coord_flip() +
  scale_fill_manual(values = c(
    "Upregulated" = col_up,
    "Downregulated" = col_down
  )) +
  labs(
    title = "Core Genes Driving Enriched GO Terms",
    x = "Gene",
    y = "Number of GO Terms"
  ) +
  theme_minimal(base_size = 14)

go_order <- go_genes_long %>%
  distinct(Description, p.adjust) %>%
  arrange(p.adjust) %>%
  pull(Description)

gene_order <- go_genes_long %>%
  dplyr::count(geneID) %>%
  arrange(desc(n)) %>%
  pull(geneID)

top_genes <- gene_order[1:25]

go_genes_pub <- go_genes_long %>%
  filter(geneID %in% top_genes) %>%
  mutate(
    Description = factor(Description, levels = rev(go_order)),
    geneID = factor(geneID, levels = top_genes)
  )

go_genes_heat_up<- ggplot(go_genes_pub,
                       aes(x = geneID, y = Description)) +
                  geom_tile(fill = "#C73E1D", color = "white", linewidth = 0.3) +
                  labs(
                    title = "Upregulated Genes Driving ER-Stress–Related GO Biological Processes",
                    subtitle = "Hypothalamus: Lactation HFHS vs Chow HFHS",
                    x = "Gene",
                    y = "GO Biological Process"
                  ) +
                  theme_minimal(base_size = 13) +
                  theme(
                    axis.text.x = element_text(
                      angle = 45,
                      hjust = 1,
                      vjust = 1,
                      size = 9
                    ),
                    axis.text.y = element_text(size = 9),
                    panel.grid = element_blank(),
                    plot.title = element_text(face = "bold", size = 15),
                    plot.subtitle = element_text(size = 11, color = "grey40")
                  )
ggsave(
  filename = "ER_stress_GO_genes_up.png",
  plot = go_genes_heat_up,
  width = 16,
  height = 8,
  dpi = 300
)
```

```{r let's look at words that appear the most in the GO terms for go_bp_up_df and go_bp_down_df in terms of frequency}

go_categories <- list(
  "ER stress / unfolded protein" = c(
    "endoplasmic reticulum",
    "unfolded",
    "protein",
    "misfolded",
    "proteostasis",
    "ER stress"
  ),
  "Lipid metabolism" = c(
    "lipid",
    "fatty",
    "cholesterol",
    "catabolic",
    "oxidation"
  ),
  "Hormone signaling" = c(
    "leptin",
    "insulin",
    "neuropeptide"
  )
)

go_plot_df <- go_up_classified %>%
  arrange(p.adjust) %>%
  slice_head(n = 220) %>%
  mutate(
    highlight = ifelse(Category == "ER stress / unfolded protein",
                       "ER stress",
                       "Other GO terms"),
    Description = factor(Description, levels = rev(Description))
  )

p <- ggplot(go_plot_df,
               aes(x = Description,
                   y = -log10(p.adjust),
                   fill = highlight)) +
          geom_col(width = 0.7) +
          coord_flip() +
          scale_fill_manual(
            values = c(
              "ER stress" = "#C73E1D",
              "Other GO terms" = "grey80"
            )
          ) +
          labs(
            title = "ER Stress GO Terms Are Enriched Among the Top Upregulated Processes",
            x = NULL,
            y = expression(-log[10]("adjusted p-value")),
            fill = NULL
          ) +
          theme_minimal(base_size = 14) +
          theme(
            panel.grid.major.y = element_blank(),
            plot.title = element_text(face = "bold")
          )

ggsave(
  filename = "ER_stress_GO_enrichment.png",
  plot = p,
  width = 16,
  height = 8,
  dpi = 300
)

ggplot(donut_df,
       aes(x = reorder(Category, n),
           y = n,
           fill = Category)) +
  geom_col(width = 0.7) +
  coord_flip() +
  labs(
    title = "Distribution of Upregulated GO Terms by Functional Category",
    subtitle = "Majority of GO terms are functionally diverse",
    x = NULL,
    y = "Number of GO terms"
  ) +
  scale_fill_manual(values = c(
    "ER stress / unfolded protein" = "#C73E1D",
    "Lipid metabolism" = "#2A7FFF",
    "Hormone signaling" = "#6A4C93",
    "Other biological processes" = "grey80"
  )) +
  theme_minimal(base_size = 14) +
  theme(legend.position = "none")

```

```{r most frequent words in the enriched GOs}
custom_stopwords <- c(
  "process", "regulation", "positive", "negative",
  "cell", "cells", "cellular",
  "response", "via", "dependent",
  "activity", "mediated",
  "signaling", "pathway", "development", "morphogenesis", "transport", "ion", "secretion", "differentiation", "transmembrane", "receptor"
)
go_up_words <- go_bp_up_df %>%
  dplyr::select(Description) %>%
  unnest_tokens(word, Description) %>%
  filter(
    !word %in% stop_words$word,
    !word %in% custom_stopwords,
    str_detect(word, "[a-z]")
  ) %>%
  dplyr::count(word, sort = TRUE)

head(go_up_words, 20)

go_down_words <- go_bp_down_df %>%
  dplyr::select(Description) %>%
  unnest_tokens(word, Description) %>%
  filter(
    !word %in% stop_words$word,
    !word %in% custom_stopwords,
    str_detect(word, "[a-z]")
  ) %>%
  dplyr::count(word, sort = TRUE)

head(go_down_words, 20)

word_GO_bp_up <- go_up_words %>%
                  slice_max(n, n = 15) %>%
                  ggplot(aes(x = reorder(word, n), y = n)) +
                  geom_col(fill = "#C73E1D", width = 0.7) +
                  coord_flip() +
                  labs(
                    title = "Most Frequent Words in Upregulated GO Terms",
                    x = NULL,
                    y = "Frequency"
                  ) +
                  theme_minimal(base_size = 14) +
                  theme(panel.grid.major.y = element_blank())

word_GO_bp_down <- go_down_words %>%
                    slice_max(n, n = 15) %>%
                    ggplot(aes(x = reorder(word, n), y = n)) +
                    geom_col(fill = "#2A7FFF", width = 0.7) +
                    coord_flip() +
                    labs(
                      title = "Most Frequent Words in Downregulated GO Terms",
                      x = NULL,
                      y = "Frequency"
                    ) +
                    theme_minimal(base_size = 14) +
                    theme(panel.grid.major.y = element_blank())
ggsave(
  filename = "ER_stress_GO_enrichmentBP_word_up.png",
  plot = word_GO_bp_up,
  width = 12,
  height = 8,
  dpi = 300
)

ggsave(
  filename = "ER_stress_GO_enrichmentBP_word_down.png",
  plot = word_GO_bp_down,
  width = 12,
  height = 8,
  dpi = 300
)
```

```{r}
library(dplyr)
library(tidyr)
library(tidytext)
library(stringr)
library(ggalluvial)

top_words_up <- go_up_words %>%
  slice_max(n, n = 5) %>%
  pull(word)

# Ensure Description is character
go_bp_up_df <- go_bp_up_df %>%
  mutate(Description = as.character(Description))

# Initialize a vector to store the assigned word
assigned_word <- character(nrow(go_bp_up_df))

for(i in seq_len(nrow(go_bp_up_df))) {
  desc <- tolower(go_bp_up_df$Description[i])
  matches <- top_words_up[str_detect(desc, fixed(top_words_up))]
  assigned_word[i] <- if(length(matches) > 0) matches[1] else NA_character_
}

# Add to dataframe and filter only matched GO terms
word_to_go_up_simple <- go_bp_up_df %>%
  mutate(word = assigned_word) %>%
  filter(!is.na(word)) %>%
  dplyr::select(word, Description) %>%
  distinct()


ggplot(word_to_go_up_simple,
       aes(axis1 = word,
           axis2 = Description,
           y = 1)) +
  geom_alluvium(aes(fill = word), width = 0.25, alpha = 0.8) +
  geom_stratum(width = 0.25, fill = "grey80", color = "black") +
  geom_text(stat = "stratum", aes(label = after_stat(stratum)), size = 3) +
  scale_x_discrete(limits = c("Word", "GO term")) +
  labs(
    title = "Top 5 GO words → GO term descriptions (Upregulated)",
    y = "Count / Flow"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    plot.title = element_text(face = "bold", hjust = 0.5),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank()
  )
```







```{r}
library(DESeq2)
library(pheatmap)
library(gridExtra)
library(RColorBrewer)
library(grid)

vsd <- vst(dds, blind = TRUE)
vsd_mat <- assay(vsd)

group_vec <- sample_info_hypo$group
names(group_vec) <- sample_info_hypo$sample_name

go_mf_df <- as.data.frame(go_mf)
top10_mf <- go_mf_df[1:10, ]

# Define 10 distinct color palettes
palettes <- list(
  colorRampPalette(brewer.pal(9, "Reds"))(100),
  colorRampPalette(brewer.pal(9, "Blues"))(100),
  colorRampPalette(brewer.pal(9, "Greens"))(100),
  colorRampPalette(brewer.pal(9, "Purples"))(100),
  colorRampPalette(brewer.pal(9, "Oranges"))(100),
  colorRampPalette(brewer.pal(9, "YlOrBr"))(100),
  colorRampPalette(brewer.pal(9, "YlGn"))(100),
  colorRampPalette(brewer.pal(9, "PuBu"))(100),
  colorRampPalette(brewer.pal(9, "RdPu"))(100),
  colorRampPalette(brewer.pal(9, "GnBu"))(100)
)

heatmaps_list <- list()

for (i in seq_len(nrow(top10_mf))) {

  term_name <- top10_mf$Description[i]
  term_genes <- unlist(strsplit(top10_mf$geneID[i], "/"))
  term_expr <- vsd_mat[rownames(vsd_mat) %in% term_genes, ]

  if (nrow(term_expr) < 2) next

  term_means <- sapply(group_order, function(g) {
    rowMeans(term_expr[ , group_vec == g, drop = FALSE])
  })
  colnames(term_means) <- group_order

  term_means_scaled <- t(scale(t(term_means)))

  # Create a pheatmap object, but capture it as a gtable
  ph <- pheatmap(
    term_means_scaled,
    cluster_rows = TRUE,
    cluster_cols = FALSE,
    fontsize_row = 6,
    main = paste0("GO MF: ", term_name),
    color = palettes[[i]],
    silent = TRUE
  )

  heatmaps_list[[i]] <- ph[[4]] # gtable object for gridExtra
}


pdf(file.path(out_dir, "Top10_GO_MF_Heatmaps.pdf"),
    width = 30, height = 16)
grid.arrange(grobs = heatmaps_list, nrow = 2)
dev.off()
```


```{r GSEA}
dds <- DESeqDataSetFromMatrix(
  countData = count_hypo_filtered,
  colData   = sample_info_hypo,
  design    = ~ group
)

dds$group <- factor(dds$group, levels = group_order)
dds <- DESeq(dds)

res_chow_lact <- results(
  dds,
  contrast = c("group", "Lactation HFHS-HFHS", "Chow-HFHS")
)

res_chow_lact <- res_chow_lact[!is.na(res_chow_lact$log2FoldChange), ]

gene_ranks <- res_chow_lact$log2FoldChange
names(gene_ranks) <- rownames(res_chow_lact)

# Sort decreasing for GSEA
gene_ranks <- sort(gene_ranks, decreasing = TRUE)

gene_df <- bitr(
  names(gene_ranks),
  fromType = "SYMBOL",
  toType   = "ENTREZID",
  OrgDb    = org.Mm.eg.db
)

# Keep only mapped genes
gene_ranks_entrez <- gene_ranks[gene_df$SYMBOL]
names(gene_ranks_entrez) <- gene_df$ENTREZID

gsea_bp <- gseGO(
  geneList     = gene_ranks_entrez,
  OrgDb        = org.Mm.eg.db,
  keyType      = "ENTREZID",
  ont          = "BP",
  minGSSize    = 10,
  maxGSSize    = 500,
  pvalueCutoff = 0.05,
  verbose      = FALSE
)

gsea_cc <- gseGO(
  geneList     = gene_ranks_entrez,
  OrgDb        = org.Mm.eg.db,
  keyType      = "ENTREZID",
  ont          = "CC",
  minGSSize    = 10,
  maxGSSize    = 500,
  pvalueCutoff = 0.05,
  verbose      = FALSE
)

bp_df <- as.data.frame(gsea_bp) %>%
  arrange(desc(abs(NES))) %>%
  slice_head(n = 20)

gg_bp <- ggplot(
  bp_df,
  aes(
    x = NES,
    y = reorder(Description, NES),
    size = -log10(p.adjust),
    color = NES
  )
) +
  geom_point() +
  scale_color_gradient2(
    low = "#2166AC",
    mid = "white",
    high = "#B2182B",
    midpoint = 0,
    name = "NES\n(<0 = Chow-HFHS\n>0 = Lactation HFHS-HFHS)"
  ) +
  labs(
    title = "GSEA – GO Biological Process",
    subtitle = "Lactation HFHS-HFHS vs Chow-HFHS",
    x = "Normalized Enrichment Score (NES)",
    y = NULL,
    size = "-log10(adj p)"
  ) +
  theme_bw(base_size = 14)

cc_df <- as.data.frame(gsea_cc) %>%
  arrange(desc(abs(NES))) %>%
  slice_head(n = 20)

gg_cc <- ggplot(
  cc_df,
  aes(
    x = NES,
    y = reorder(Description, NES),
    size = -log10(p.adjust),
    color = NES
  )
) +
  geom_point() +
  scale_color_gradient2(
    low = "#2166AC",
    mid = "white",
    high = "#B2182B",
    midpoint = 0,
    name = "NES\n(<0 = Chow-HFHS\n>0 = Lactation HFHS-HFHS)"
  ) +
  labs(
    title = "GSEA – GO Cellular Component",
    subtitle = "Lactation HFHS-HFHS vs Chow-HFHS",
    x = "Normalized Enrichment Score (NES)",
    y = NULL,
    size = "-log10(adj p)"
  ) +
  theme_bw(base_size = 14)

ggsave(
  filename = file.path(out_dir, "GO_BP_dotplot.png"),
  plot = gg_bp,
  width = 9,
  height = 8,
  dpi = 300
)

ggsave(
  filename = file.path(out_dir, "GO_MF_dotplot.png"),
  plot = gg_cc,
  width = 9,
  height = 8,
  dpi = 300
)

```


```{r leptin signaling pathway}
go_leptin <- keggGet("GO:0044321")[[1]]$GENE

# Extract Entrez IDs only (even indices)
entrez_ids <- go_leptin[seq(1, length(kegg_leptin), 2)]

# Map Entrez IDs to gene symbols
leptin_genes <- bitr(
  entrez_ids,
  fromType = "ENTREZID",
  toType   = "SYMBOL",
  OrgDb    = org.Mm.eg.db
)$SYMBOL

# VST-normalized expression
vsd <- vst(dds, blind = TRUE)
vsd_mat <- assay(vsd)

# Subset to leptin pathway genes
leptin_expr <- vsd_mat[rownames(vsd_mat) %in% leptin_genes, ]
expr_df <- as.data.frame(leptin_expr)
expr_df$Gene <- rownames(expr_df)

expr_long <- expr_df %>%
  tidyr::pivot_longer(-Gene, names_to = "Sample", values_to = "Expression") %>%
  left_join(sample_info_hypo[, c("sample_name", "group")],
            by = c("Sample" = "sample_name"))

expr_avg <- expr_long %>%
  group_by(Gene, group) %>%
  summarise(mean_expr = mean(Expression, na.rm = TRUE), .groups = "drop")

expr_matrix <- expr_avg %>%
  tidyr::pivot_wider(names_from = group, values_from = mean_expr) %>%
  column_to_rownames("Gene") %>%
  as.matrix()
expr_scaled <- t(scale(t(expr_matrix)))

annotation_col <- data.frame(group = colnames(expr_scaled))
rownames(annotation_col) <- colnames(expr_scaled)
ann_colors <- list(group = group_colors)

pdf(
  file.path(out_dir, "KEGG_Leptin Signaling Genes Heatmap.pdf"),
  width = 8,
  height = 9
)

pheatmap(
  expr_scaled,
  cluster_rows = TRUE,
  cluster_cols = FALSE,  # already averaged by group
  annotation_col = annotation_col,
  #annotation_colors = ann_colors,
  fontsize_row = 9,
  main = "Leptin Signaling Genes – Mean Expression per Group"
)

dev.off()
```

```{r negative regulation of insulin receptor signaling}

go_leptin <- as.list(org.Hs.egGO2ALLEGS[["GO:0046627"]])  # ENTREZ IDs
go_leptin <- unlist(go_leptin)

# Map Entrez IDs to gene symbols
leptin_genes <- bitr(
  go_leptin,
  fromType = "ENTREZID",
  toType   = "SYMBOL",
  OrgDb    = org.Hs.eg.db
)$SYMBOL

# --- 2. VST-normalized expression ---
vsd <- vst(dds, blind = TRUE)
vsd_mat <- assay(vsd)

# Make uppercase copies for matching
vsd_genes_upper <- toupper(rownames(vsd_mat))
leptin_genes_upper <- toupper(leptin_genes)

# Subset using case-insensitive match
leptin_expr <- vsd_mat[vsd_genes_upper %in% leptin_genes_upper, ]

# Keep original gene symbols
expr_df <- as.data.frame(leptin_expr)
expr_df$Gene <- rownames(expr_df)

# --- 3. Convert to long format and join sample info ---
expr_long <- expr_df %>%
  pivot_longer(-Gene, names_to = "Sample", values_to = "Expression") %>%
  left_join(sample_info_hypo[, c("sample_name", "group")],
            by = c("Sample" = "sample_name"))

# --- 4. Average by group ---
expr_avg <- expr_long %>%
  group_by(Gene, group) %>%
  summarise(mean_expr = mean(Expression, na.rm = TRUE), .groups = "drop")

expr_matrix <- expr_avg %>%
  pivot_wider(names_from = group, values_from = mean_expr) %>%
  column_to_rownames("Gene") %>%
  as.matrix()

# --- 5. Scale rows (genes) for heatmap ---
expr_scaled <- t(scale(t(expr_matrix)))  # scale by row

# --- 6. Column annotation ---
annotation_col <- data.frame(Group = colnames(expr_scaled))
rownames(annotation_col) <- colnames(expr_scaled)
ann_colors <- list(Group = group_colors)  # define group_colors for your groups

# --- 7. Plot heatmap ---
pdf(file.path(out_dir, "negative regulation of insulin receptor signaling heatmap.pdf"), width = 8, height = 9)

pheatmap(
  expr_scaled,
  cluster_rows = TRUE,
  cluster_cols = FALSE,  # already averaged per group
  annotation_colors = ann_colors,
  fontsize_row = 9,
  fontsize_col = 10,
  main = "negative regulation of insulin receptor signaling",
  angle_col = 45,       # rotate x-axis labels
  border_color = "grey60",
  color = colorRampPalette(c("navy", "white", "firebrick3"))(50)
)

dev.off()

```

```{r positive regulation of insulin receptor signaling}

go_leptin <- as.list(org.Hs.egGO2ALLEGS[["GO:0046628"]])  # ENTREZ IDs
go_leptin <- unlist(go_leptin)

# Map Entrez IDs to gene symbols
leptin_genes <- bitr(
  go_leptin,
  fromType = "ENTREZID",
  toType   = "SYMBOL",
  OrgDb    = org.Hs.eg.db
)$SYMBOL

# --- 2. VST-normalized expression ---
vsd <- vst(dds, blind = TRUE)
vsd_mat <- assay(vsd)

# Make uppercase copies for matching
vsd_genes_upper <- toupper(rownames(vsd_mat))
leptin_genes_upper <- toupper(leptin_genes)

# Subset using case-insensitive match
leptin_expr <- vsd_mat[vsd_genes_upper %in% leptin_genes_upper, ]

# Keep original gene symbols
expr_df <- as.data.frame(leptin_expr)
expr_df$Gene <- rownames(expr_df)

# --- 3. Convert to long format and join sample info ---
expr_long <- expr_df %>%
  pivot_longer(-Gene, names_to = "Sample", values_to = "Expression") %>%
  left_join(sample_info_hypo[, c("sample_name", "group")],
            by = c("Sample" = "sample_name"))

# --- 4. Average by group ---
expr_avg <- expr_long %>%
  group_by(Gene, group) %>%
  summarise(mean_expr = mean(Expression, na.rm = TRUE), .groups = "drop")

expr_matrix <- expr_avg %>%
  pivot_wider(names_from = group, values_from = mean_expr) %>%
  column_to_rownames("Gene") %>%
  as.matrix()

# --- 5. Scale rows (genes) for heatmap ---
expr_scaled <- t(scale(t(expr_matrix)))  # scale by row

# --- 6. Column annotation ---
annotation_col <- data.frame(Group = colnames(expr_scaled))
rownames(annotation_col) <- colnames(expr_scaled)
ann_colors <- list(Group = group_colors)  # define group_colors for your groups

# --- 7. Plot heatmap ---
pdf(file.path(out_dir, "positive regulation of insulin receptor signaling heatmap.pdf"), width = 8, height = 9)

pheatmap(
  expr_scaled,
  cluster_rows = TRUE,
  cluster_cols = FALSE,  # already averaged per group
  annotation_colors = ann_colors,
  fontsize_row = 9,
  fontsize_col = 10,
  main = "positive regulation of insulin receptor signaling",
  angle_col = 45,       # rotate x-axis labels
  border_color = "grey60",
  color = colorRampPalette(c("navy", "white", "firebrick3"))(50)
)

dev.off()

```

```{r gene of interest}
genes_of_interest <- c("Irs2", "Socs3", "Ptpn1", "Stat3", "Ucp2", "Mfn1", "Sim1", "Bdnf", "Ntrk2", "Xbp1", "Crhr2","Foxo1", "Pik3ca", "Nrp1", "Nrp2", "Sh2b1", "Pcsk1", "Phip", "Oxtr", "Hspa5")
count_goi <- count_hypo_filtered[
  rownames(count_hypo_filtered) %in% genes_of_interest,
  sample_cols,
  drop = FALSE
]

sample_info_hypo <- sample_info_hypo[
  match(colnames(count_goi), sample_info_hypo$sample_name),
]

group_means <- sapply(group_order, function(g) {
  samples_in_group <- sample_info_hypo$sample_name[
    sample_info_hypo$group == g
  ]
  
  rowMeans(count_goi[ , samples_in_group, drop = FALSE])
})

# convert to matrix
group_means <- as.matrix(group_means)

group_means_scaled <- t(scale(t(group_means)))

pheatmap(
  group_means_scaled,
  color = colorRampPalette(c("navy", "white", "firebrick3"))(100),
  cluster_rows = TRUE,
  cluster_cols = FALSE,
  show_colnames = TRUE,
  show_rownames = TRUE,
  fontsize_row = 10,
  fontsize_col = 10,
  main = "Mean Expression per Group"
)
```


```{r Leptin-Melanocortin Pathway}
genes_leptin_melanocortin <- c("Lepr", "Mc4r", "Pcsk1", "Pomc", "Adcy3", "Alms1", "Arnt2", "Bdnf", "Cep19", "Krs2", "Magel2", "Mc3r", "Mrap2", "Myt1l", "Ncoa1", "Ntrk2", "Sh2b1", "Sim1", "Tub")
count_goi <- count_hypo_filtered[
  rownames(count_hypo_filtered) %in% genes_leptin_melanocortin,
  sample_cols,
  drop = FALSE
]

sample_info_hypo <- sample_info_hypo[
  match(colnames(count_goi), sample_info_hypo$sample_name),
]

group_means <- sapply(group_order, function(g) {
  samples_in_group <- sample_info_hypo$sample_name[
    sample_info_hypo$group == g
  ]
  
  rowMeans(count_goi[ , samples_in_group, drop = FALSE])
})

# convert to matrix
group_means <- as.matrix(group_means)

scaled_matrix <- t(scale(t(group_means)))

# Colors
my_colors <- colorRampPalette(c("blue", "white", "red"))(100)

pheatmap(
  scaled_matrix,
  cluster_rows = TRUE,       # cluster genes
  cluster_cols = FALSE,      # keep group order fixed
  color = my_colors,
  border_color = NA,
  fontsize_row = 10,
  fontsize_col = 12,
  main = "Leptin-Melanocortin Gene Expression Across Groups",
  angle_col = 45
)

# Melt matrix for ggplot
df_plot <- melt(group_means)
colnames(df_plot) <- c("Gene", "Group", "Expression")

ggplot(df_plot, aes(x = Gene, y = Expression, fill = Group)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme_minimal(base_size = 14) +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Average Expression of Leptin-Melanocortin Genes",
       y = "Expression (counts)", x = "") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Compute mean per group
composite_score <- colMeans(group_means)

# Put into dataframe
df_score <- data.frame(
  Group = names(composite_score),
  Score = composite_score
)

ggplot(df_score, aes(x = Group, y = Score, fill = Group)) +
  geom_col() +
  geom_text(aes(label = round(Score, 2)), vjust = -0.5) +
  theme_minimal(base_size = 14) +
  scale_fill_brewer(palette = "Set2") +
  labs(title = "Melanocortin Pathway Composite Score per Group",
       y = "Mean Expression", x = "")
```